gpytorch/utils/minres.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+
+import torch
+
+from .. import settings
+from .broadcasting import _pad_with_singletons
+
+
+def minres(matmul_closure, rhs, eps=1e-25, shifts=None, value=None, max_iter=None, preconditioner=None):
+    r"""
+    Perform MINRES to find solutions to :math:`(\mathbf K + \alpha \sigma \mathbf I) \mathbf x = \mathbf b`.
+    Will find solutions for multiple shifts :math:`\sigma` at the same time.
+
+    :param callable matmul_closure: Function to perform matmul with.
+    :param torch.Tensor rhs: The vector :math:`\mathbf b` to solve against.
+    :param torch.Tensor shifts: (default None) The shift :math:`\sigma` values. If set to None,
+        then :math:`\sigma=0`.
+    :param float value: (default None) The multiplicative constant :math:`\alpha`. If set to None,
+        then :math:`\alpha=0`.
+    :param int max_iter: (default None) The maximum number of minres iterations. If set to None, then
+        uses the constant stored in :obj:`gpytorch.settings.max_cg_iterations`.
+    :rtype: torch.Tensor
+    :return: The solves :math:`\mathbf x`. The shape will correspond to the size of `rhs` and `shifts`.
+    """
+    # Default values
+    if torch.is_tensor(matmul_closure):
+        matmul_closure = matmul_closure.matmul
+    mm_ = matmul_closure
+    if preconditioner is None:
+        preconditioner = lambda x: x.clone()
```

Comment:
why clone?

Reply:
We later modify the output of the preconditioner in place:
https://github.com/cornellius-gp/gpytorch/blob/fac03c2622e72288e2edf5d5b3edce7cbc7b3846/gpytorch/utils/minres.py#L75